util/src/main/java/org/killbill/billing/util/entity/dao/EntitySqlDao.java

```diff
@@ -31,23 +31,27 @@
 import org.killbill.commons.jdbi.statement.SmartFetchSize;
 import org.killbill.commons.jdbi.template.KillBillSqlDaoStringTemplate;
 import org.skife.jdbi.v2.sqlobject.Bind;
+import org.skife.jdbi.v2.sqlobject.GetGeneratedKeys;
 import org.skife.jdbi.v2.sqlobject.SqlBatch;
 import org.skife.jdbi.v2.sqlobject.SqlQuery;
 import org.skife.jdbi.v2.sqlobject.SqlUpdate;
 import org.skife.jdbi.v2.sqlobject.customizers.BatchChunkSize;
 import org.skife.jdbi.v2.sqlobject.customizers.Define;
 import org.skife.jdbi.v2.sqlobject.mixins.CloseMe;
 import org.skife.jdbi.v2.sqlobject.mixins.Transactional;
+import org.skife.jdbi.v2.util.LongMapper;
 
 @KillBillSqlDaoStringTemplate
 public interface EntitySqlDao<M extends EntityModelDao<E>, E extends Entity> extends AuditSqlDao, HistorySqlDao<M, E>, Transactional<EntitySqlDao<M, E>>, CloseMe {
 
     @SqlUpdate
+    @GetGeneratedKeys(value = LongMapper.class)
     @Audited(ChangeType.INSERT)
     public Object create(@SmartBindBean final M entity,
                          @SmartBindBean final InternalCallContext context);
 
     @SqlBatch
+    // We don't @GetGeneratedKeys here, as it's unclear if the ordering through the batches is respected by all JDBC drivers
```

Comment:
The JDBI 3 (but 2 is probably no different) [seems to indicate we can use it](http://jdbi.org/#__getgeneratedkeys_3). 

I suppose in the current scheme, the code follows with a get to retrieve all objects and access the auto-generated fields?

Reply:
> The JDBI 3 (but 2 is probably no different) [seems to indicate we can use it](http://jdbi.org/#__getgeneratedkeys_3).

Our version of JDBI doesn't support `@GetGeneratedKeys` on batch statements, but I'm currently fixing it.

But the remark here was more about the **order** of keys returned: it is driver specific whether the ordering returned matches the ordering of inserts.

> I suppose in the current scheme, the code follows with a get to retrieve all objects and access the auto-generated fields?

Yes. Note also that for `INSERT` with a history table, we need the full rehydrated object anyways.